src/resolution/common.rs

```diff
@@ -0,0 +1,248 @@
+      Some(v) => Ok(v),
+      // If the package isn't found, it likely means that the user needs to use
+      // `--reload` in the CLI to get the latest npm package information. Although
+      // it seems like we could make this smart by fetching the latest information for
+      // this package here, we really need a full restart. There could be very
+      // interesting bugs that occur if this package's version was resolved by
+      // something previous using the old information, then now being smart here
+      // causes a new fetch of the package information, meaning this time the
+      // previous resolution of this package's version resolved to an older
+      // version, but next time to a different version because it has new information.
+      None => Err(NpmPackageVersionResolutionError::VersionReqNotMatched {
+        package_name: info.name.clone(),
+        version_req: version_req.clone(),
+      }),
+    }
+  }
+}
+
+pub fn version_req_satisfies(
+  version_req: &VersionReq,
+  version: &Version,
+  package_info: &NpmPackageInfo,
+) -> Result<bool, NpmPackageVersionResolutionError> {
+  match version_req.tag() {
+    Some(tag) => {
+      let version_info = tag_to_version_info(package_info, tag)?;
+      Ok(version_info.version == *version)
+    }
+    None => Ok(version_req.matches(version)),
+  }
+}
+
+fn resolve_best_from_existing_versions<'a>(
+  version_req: &VersionReq,
+  package_info: &NpmPackageInfo,
+  existing_versions: impl Iterator<Item = &'a Version>,
+) -> Result<Option<&'a Version>, NpmPackageVersionResolutionError> {
+  let mut maybe_best_version: Option<&Version> = None;
+  for version in existing_versions {
+    if version_req_satisfies(version_req, version, package_info)? {
+      let is_best_version = maybe_best_version
+        .as_ref()
+        .map(|best_version| (*best_version).cmp(version).is_lt())
+        .unwrap_or(true);
+      if is_best_version {
+        maybe_best_version = Some(version);
+      }
+    }
+  }
+  Ok(maybe_best_version)
+}
+
+fn tag_to_version_info<'a>(
+  info: &'a NpmPackageInfo,
+  tag: &str,
+) -> Result<&'a NpmPackageVersionInfo, NpmPackageVersionResolutionError> {
+  // For when someone just specifies @types/node, we want to pull in a
+  // "known good" version of @types/node that works well with Deno and
+  // not necessarily the latest version. For example, we might only be
+  // compatible with Node vX, but then Node vY is published so we wouldn't
+  // want to pull that in.
+  // Note: If the user doesn't want this behavior, then they can specify an
+  // explicit version.
+  if tag == "latest" && info.name == "@types/node" {
+    return get_resolved_package_version_and_info(
+      // todo(dsherret): follow up and remove this in a future PR
+      // WARNING: When bumping this version, check if anything needs to be
+      // updated in the `setNodeOnlyGlobalNames` call in 99_main_compiler.js
+      &VersionReq::parse_from_npm("18.0.0 - 18.11.18").unwrap(),
```

Comment:
This could be abstracted to a `const TYPES_NODE` in a separate module so we could bump it using a script

Reply:
See the todo above. I'm going to remove it.